### PR TITLE
[Frontend] Cap number of metadata cache entries

### PR DESCRIPTION
**Pull request by snalli**
This PR adds an ability to restrict the maximum number of entries
in the metadata cache. This allows us to control the amount of memory
consumed by the metadata cache.

### CODE REVIEW

**Comment by snalli** [comment]
This is my last patch for caching.

**Review by Arun-LinkedIn — APPROVED** [review]

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 1]
```diff
@@ -1411,6 +1411,108 @@ public void testRouterMetadataCacheSimpleBlob() throws Exception {
...
+      int TEST_MAX_NUM_METADATA_CACHE_ENTRIES = 1;
+      int numMisses = NUM_GET_COMPOSITE_BLOBS;
+      int numInserts = NUM_GET_COMPOSITE_BLOBS;
+      Properties properties = getNonBlockingRouterProperties("DC1");
+      properties.setProperty("router.blob.metadata.cache.enabled", Boolean.toString(true));
+      properties.setProperty("router.smallest.blob.for.metadata.cache", Long.toString(0));
+      // Set the size of cache to 1 entry which means the cache can store at max of one entry at any time
+      properties.setProperty(RouterConfig.ROUTER_MAX_NUM_METADATA_CACHE_ENTRIES, Integer.toString(TEST_MAX_NUM_METADATA_CACHE_ENTRIES));
+      setRouterWithMetadataCache(properties, new AmbryCacheMetrics(0, numMisses, numInserts,0));
+      AmbryCacheWithMetrics ambryCacheWithMetrics = (AmbryCacheWithMetrics) router.getBlobMetadataCache();
```
Comment (justinlin-linkedin) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1432]: it's AmbryCacheWithStats now.

**Review by justinlin-linkedin — CHANGES_REQUESTED** [review]

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java [hunk 2]
```diff
@@ -608,13 +608,20 @@ public class RouterConfig {
   public final long routerSmallestBlobForMetadataCache;
   public static final long NUM_BYTES_IN_ONE_TB = (long) Math.pow(1024, 4);
 
+  public static final String ROUTER_MAX_NUM_METADATA_CACHE_ENTRIES = "router.max.num.metadata.cache.entries";
+  @Config(ROUTER_MAX_NUM_METADATA_CACHE_ENTRIES)
+  public final int routerMaxNumMetadataCacheEntries;
+  public static final int MAX_NUM_METADATA_CACHE_ENTRIES = 10;
```
Comment (panchow) [inline ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:614]: Let's make the name consistent.  Suggest changing name to MAX_NUM_METADATA_CACHE_ENTRIES_DEFAULT

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 3]
```diff
@@ -1411,6 +1411,108 @@ public void testRouterMetadataCacheSimpleBlob() throws Exception {
...
+        // get the one byte from composite blob, and it's metadata must be present in cache
+        ByteRange range = ByteRanges.fromOffsetRange(1, 2);
+        String blobId = blobIds.get(i % NUM_PUT_COMPOSITE_BLOBS);
+        router.getBlob(blobId, new GetBlobOptionsBuilder().range(range).build()).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+      }
+
+      assertTrue(String.format("Must encounter %s cache misses", numMisses), ambryCacheWithMetrics.getCacheMissCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertTrue(String.format("Must encounter %s cache inserts", numInserts), ambryCacheWithMetrics.getPutObjectCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    } finally {
```
Comment (panchow) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1459]: I guess this is a mocked router.  It is created once but closed multiple times (one in each test).  If it's a mock, don't have to do anything about it.

**Review by panchow — APPROVED** [review]

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 1 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1432]: Thanks will fix.

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java (hunk 2 above)
Comment (snalli) [reply ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:614]: thanks

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 3 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1459]: for each test, the entire test class is instantiated, router created, test run and then router closed.

File: ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java [hunk 4]
```diff
@@ -165,7 +171,7 @@ public boolean deleteObject(String key) {
   /**
    * Initialize metrics for this instance of AmbryCache
    */
-  private void initializeAmbryCacheMetrics() {
+  private void initializeAmbryCacheStats() {
```
Comment (justinlin-linkedin) [inline ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java:174]: this should be metrics, because you are initializing metrics.

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java (hunk 4 above)
Comment (snalli) [reply ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java:174]: just saw this comment, will fix

**Pull request merged**